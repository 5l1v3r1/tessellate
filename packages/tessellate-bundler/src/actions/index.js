// @flow

import * as bundles from './bundles'
import * as content from './content'
