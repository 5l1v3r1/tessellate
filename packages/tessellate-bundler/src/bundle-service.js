// @flow
import webpack from 'webpack'
import ExtractTextPlugin from 'extract-text-webpack-plugin'
import WebpackSandbox from 'webpack-sandboxed'
import logger from './logger'
import * as utils from './utils'

type Options = {
  cssSupport?: boolean;
  production?: boolean;
  packages?: Array<string>;
  externals?: { [key: string]: string };
}

const log = logger('bundle-service')

async function _createWebpackSandbox(args: Options = {}): Promise<WebpackSandbox> {
  const loaders = [{
    test: /\.js$/,
    exclude: /node_modules/,
    loader: 'babel-loader',
    options: {
      presets: [['es2015', {modules: false}], 'react'],
      babelrc: false
    }
  }]
  const plugins = [
    new webpack.DefinePlugin({process: {env: {NODE_ENV: '"production"'}}})
  ]
  const externals = Object.assign({}, args.externals, {
    'react': 'react',
    'react-dom': 'react-dom'
  })
  const packages = args.packages || []

  if (args.cssSupport) {
    loaders.push({
      test: /\.css$/, loader: ExtractTextPlugin.extract({
        fallbackLoader: 'style-loader',
        loader: 'css-loader'
      })
    })
    plugins.push(
      new ExtractTextPlugin('[name]-[hash].min.css')
    )
    packages.push(
      'style-loader',
      'css-loader'
    )
  }

  if (args.production) {
    plugins.push(
      new webpack.optimize.UglifyJsPlugin(),
      new webpack.optimize.DedupePlugin()
    )
  }

  return WebpackSandbox.createInstance({
    config: {
      target: 'web',
      module: {
        loaders
      },
      output: {
        libraryTarget: 'umd'
      },
      plugins,
      externals
    },
    packages
  })
}

const createWebpackSandbox = utils.memoize(_createWebpackSandbox)

export async function make(source: string, options: Options = {}): Promise<TessellateBundle> {
  log.debug('Create WebpackSandbox with options %o', options)
  const sandbox = await createWebpackSandbox(options)
  log.debug('Compile bundle...')
  const [bundle] = await sandbox.run(source)

  const files = Object.keys(bundle)
  log.debug('Made bundle %o', files)
  // TODO: handle case of multiple files
  const jsBundle = files.filter(b => b.endsWith('.js'))[0]
  const cssBundle = files.filter(b => b.endsWith('.css'))[0]

  const js = {
    name: jsBundle,
    source: bundle[jsBundle]
  }
  const css = cssBundle ? {
    name: cssBundle,
    source: bundle[cssBundle]
  } : undefined

  return {js, css}
}
