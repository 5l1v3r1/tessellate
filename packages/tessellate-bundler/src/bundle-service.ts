import ExtractTextPlugin = require('extract-text-webpack-plugin');
import { log } from 'tessellate-server';
import webpack = require('webpack');
import webpackSandboxed, { WebpackRunner } from 'webpack-sandboxed';
import TessellateBundle from './TessellateBundle';
import * as utils from './utils';

export type Options = {
  cssSupport?: boolean;
  production?: boolean;
  packages?: string[];
  externals?: { [key: string]: string };
};

async function _createWebpackSandbox(args: Options = {}): Promise<WebpackRunner> {
  const rules: webpack.Rule[] = [
    {
      test: /\.js$/,
      exclude: /node_modules/,
      loader: 'babel-loader',
      options: {
        presets: ['es2015', 'react'],
        babelrc: false
      }
    }
  ];

  const plugins: webpack.Plugin[] = [
    new webpack.DefinePlugin({ 'process.env.NODE_ENV': JSON.stringify('production') })
  ];

  const externals: webpack.ExternalsObjectElement = Object.assign({}, args.externals, {
    'react': 'react',
    'react-dom': 'react-dom'
  });

  const packages = args.packages || [];
  packages.push('babel-loader');

  if (args.cssSupport) {
    rules.push({
      test: /\.css$/,
      use: ExtractTextPlugin.extract({
        fallback: 'style-loader',
        use: 'css-loader'
      })
    });
    plugins.push(new ExtractTextPlugin('[name]-[hash].min.css'));
    packages.push('style-loader', 'css-loader', 'extract-text-webpack-plugin');
  }

  if (args.production) {
    plugins.push(new webpack.optimize.UglifyJsPlugin(), new webpack.optimize.DedupePlugin());
  }

  return webpackSandboxed({
    config: {
      target: 'web',
      module: {
        rules
      },
      output: {
        libraryTarget: 'umd'
      },
      plugins,
      externals
    },
    packages
  });
}

const createWebpackSandbox = utils.memoize(_createWebpackSandbox);

export async function make(source: string, options: Options = {}): Promise<TessellateBundle> {
  log.info('Create WebpackSandbox with options %j', options);
  const webpackSandbox = await createWebpackSandbox(options);
  const [bundle] = await webpackSandbox.run(source);

  const files = Object.keys(bundle);
  log.info('Made bundle %j', files);
  // TODO: handle case of multiple files
  const jsBundle = files.filter(b => b.endsWith('.js'))[0];
  const cssBundle = files.filter(b => b.endsWith('.css'))[0];

  const js = {
    name: jsBundle,
    source: bundle[jsBundle]
  };
  const css = cssBundle
    ? {
      name: cssBundle,
      source: bundle[cssBundle]
    }
    : undefined;

  return { js, css };
}
