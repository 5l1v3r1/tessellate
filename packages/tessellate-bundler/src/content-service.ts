import url = require('url');
import path = require('path');
import fs = require('mz/fs');
import thenify = require('thenify');
import { conf, log, Problem } from 'tessellate-server';
import TessellateBundle from './TessellateBundle';
import TessellateElement from './TessellateElement';

export type ExportArgs = {
  domain: string;
  name: string;
  bundle: TessellateBundle;
};

export type FetchArgs = {
  domain: string;
  name: string;
};

export type PublishResult = {
  js: string;
  css?: string;
};

class ContentProblem extends Problem {
  constructor(detail: string) {
    super({ title: 'Content error.', detail, status: 400 });
  }
}

const mkdirp = thenify(require('mkdirp'));

export async function publish(args: ExportArgs): Promise<PublishResult> {
  log.debug('Publish %s/%s', args.domain, args.name);
  const target = url.parse(conf.getString('publishTarget', ''));

  if (target.protocol === 'file:') {
    let targetPath;
    if (target.hostname && target.pathname) {
      targetPath = path.join(target.hostname, target.pathname);
    } else if (target.pathname) {
      targetPath = target.pathname;
    } else {
      throw new ContentProblem('Undefined target path.');
    }
    return await publishToFileSystem(targetPath, args);
  } else if (target.protocol === 's3:') {
    return publishToS3(target.hostname, target.pathname);
  } else {
    throw new ContentProblem(`Illegal target protocol ${target.protocol || '???'}`);
  }
}

async function publishToFileSystem(target: string, args: ExportArgs): Promise<PublishResult> {
  const basePath = path.join(target, args.domain, args.name);
  log.debug(`Write bundle to ${basePath}`);

  const jsPath = path.join(basePath, 'index.js');
  const cssPath = path.join(basePath, 'index.css');
  const result: PublishResult = { js: path.relative(target, jsPath), css: undefined };

  await mkdirp(basePath);
  await fs.writeFile(path.resolve(process.cwd(), jsPath), args.bundle.js.source);

  if (typeof args.bundle.css === 'object' && typeof args.bundle.css.source === 'string') {
    const cssSource = args.bundle.css.source;
    await fs.writeFile(path.resolve(process.cwd(), cssPath), cssSource);
    result.css = path.relative(target, cssPath);
  }

  return result;
}

async function publishToS3(..._: any[]): Promise<PublishResult> {
  throw new Error('publishToS3 not implemented.');
}

export async function fetch(args: FetchArgs): Promise<TessellateElement> {
  log.debug('Fetch %s/%s', args.domain, args.name);
  throw new Error('fetch not implemented.');
}
