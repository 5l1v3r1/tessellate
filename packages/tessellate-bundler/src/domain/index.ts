export { default as Bundler } from './Bundler';
export { default as BundleService } from './BundleService';
export { default as HealthService } from './HealthService';
export { default as WebpackFactory } from './WebpackFactory';
