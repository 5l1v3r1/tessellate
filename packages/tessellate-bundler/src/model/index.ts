export { File } from './TessellateBundle';
export { TessellateBundle } from './TessellateBundle';
export { TessellateElement } from './TessellateElement';
