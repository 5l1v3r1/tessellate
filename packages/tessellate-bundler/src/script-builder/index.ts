import uuid = require('uuid');
import { camelCase } from 'change-case';
import { log, Problem } from 'tessellate-server';
import TessellateElement from '../TessellateElement';
import createElementScript from './CreateElementScript';
import fragmentScript from './FragmentScript';

export type ReactScript = {
  script: string;
  props: { [key: string]: any };
  imports: { [key: string]: any };
};

type PropsType = { [key: string]: any };
type ImportsType = { [key: string]: (string[] | string) };

class ElementProblem extends Problem {
  constructor(detail: string) {
    super({ title: 'TessellateElement error.', detail, status: 400 });
  }
}

function processChildren(children: ReactScript[]) {
  return children.reduce(
    ({ childScripts, childProps }, { script, props }) => ({
      childScripts: childScripts.concat(script),
      childProps: Object.assign(childProps, props)
    }),
    { childScripts: [] as string[], childProps: {} }
  );
}

export function build(element: TessellateElement): string {
  log.debug('Build fragment script...');
  const { script, props, imports } = toReactScript(element);
  return fragmentScript({ reactElement: script, rootID: uuid.v4(), props, imports });
}

function toReactScript(
  element: TessellateElement | string,
  props: PropsType = {},
  imports: ImportsType = {}
): ReactScript {
  // Atomic element, e.g. string content of a headline.
  if (typeof element === 'string') return { script: `\`${element}\``, props, imports };

  const { className } = parseElementType(element, imports);
  const id = uuid.v4();
  props[id] = element.props;

  const children = toReactScripts(element.children, props, imports);
  const { childScripts, childProps } = processChildren(children);

  const script = createElementScript({ className, propsId: id, children: childScripts });

  return { script, props: Object.assign(props, childProps), imports };
}

function toReactScripts(
  elements: Array<TessellateElement | string> | string | undefined,
  props: PropsType,
  imports: ImportsType
): ReactScript[] {
  if (typeof elements === 'string') return [toReactScript(elements, props, imports)];
  if (Array.isArray(elements)) return elements.map(child => toReactScript(child, props, imports));
  else return [];
}

function parseElementType(
  element: TessellateElement,
  imports: ImportsType
): { className: string, imports: ImportsType } {
  if (!element.type) throw new ElementProblem('Missing element type on ' + JSON.stringify(element));

  // Try to match '<node-module-name>.<component-name>'
  const [moduleName, componentName] = (element.type.match(/^([^\.]+)\.?(.+)?/) || []).slice(1);

  if (moduleName && componentName) {
    // If a style attribute is present, add a module import for it.

    if (element.style) {
      const styleImport = camelCase(element.style);
      imports[element.style] = styleImport;
    }

    // Add a module import based on the type.
    let className;
    // Them import is 'import <import-name> from <node-module-name>'
    if (componentName === 'default') {
      const importName: string = camelCase(moduleName);
      imports[moduleName] = importName;
      className = importName;
    } else {
      // The import is 'import { <component-name> } from '<node-module-name>'
      className = componentName;
      const [importName] = (componentName.match(/^([^\.]+)\.?(.+)?/) || []).slice(1);
      const imported = imports[moduleName];
      if (Array.isArray(imported)) {
        if (!imports[moduleName].includes(importName)) {
          imports[moduleName] = imported.concat(importName);
        }
      } else {
        imports[moduleName] = [importName];
      }
    }

    return { className, imports };
  } else if (moduleName) {
    // Type is a literal React component name (e.g. 'div')
    const name: string = camelCase(moduleName);
    const className = `'${name}'`;
    return { className, imports };
  } else throw new ElementProblem(`Illegal element type ${element.type}`);
}
