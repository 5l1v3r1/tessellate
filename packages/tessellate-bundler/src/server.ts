/// <reference path="../typings/index.d.ts" />

import url = require('url');
import path = require('path');
import koaStatic = require('koa-static');
import kcors = require('kcors');
import { conf, log, TessellateServer } from 'tessellate-server';
import bundleEpic from './epics/bundles';

function getFileSystemPublishTarget(): string | null {
  const target = url.parse(conf.getString('PUBLISH_TARGET'));
  if (target.protocol === 'file:') {
    const { hostname, pathname } = target;
    if (hostname || pathname) {
      return path.resolve(process.cwd(), path.join(hostname || '', pathname || ''));
    } else {
      return null;
    }
  }
  return null;
}

export function init(): TessellateServer {
  const server = new TessellateServer();

  server.use(kcors());

  const fileSystemPublishTarget = getFileSystemPublishTarget();
  if (fileSystemPublishTarget) {
    log.info('Serving files from %s', fileSystemPublishTarget);
    server.use(koaStatic(fileSystemPublishTarget, { defer: true, gzip: true }));
  }

  server.router.get('/health', ctx => ctx.body = 'OK');
  server.router.post('/bundles/:domain/:name', bundleEpic);

  return server;
}

export async function start(
  port = conf.getNumber('APP_PORT')
): Promise<TessellateServer> {
  const server = await init().start(port, port + 1);
  log.info('listening on port %d', port);
  return server;
}

export function main() {
  start().catch(log.error);
}

if (require.main === module) {
  main();
}
