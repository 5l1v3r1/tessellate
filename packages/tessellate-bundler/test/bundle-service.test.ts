/// <reference types="jest" />

jest.mock('mkdirp', () => (_: string, callback: () => void) => callback());
require('mz/fs').writeFile = jest.fn(() => Promise.resolve());

import fs = require('mz/fs');
import path = require('path');
import createWebpackSandbox from 'webpack-sandboxed';
import * as bundleService from '../src/bundle-service';

describe('bundle-service', () => {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

  beforeEach(() => (jest.resetAllMocks()));

  test('ensure webpack-sandboxed works', async () => {
    const webpackSandbox = await createWebpackSandbox();
    const [bundle] = await webpackSandbox.run('module.exports = 42;');
    expect(Object.keys(bundle).length).toBe(1);
    expect(Object.keys(bundle)[0]).toMatch(/^main\.\S+\.js$/);
  });

  test('make a bundle from a source string', async () => {
    const source = await fs.readFile(path.resolve(__dirname, 'fixtures', 'fragment-script.js'));
    const bundle = await bundleService.make(source.toString());
    expect(bundle.js).toBeDefined();
    expect(bundle.css).not.toBeDefined();
    expect(bundle.js.name).toMatch(/\.js$/);
    expect(bundle.js.source.toString()).toMatchSnapshot();
  });
});
