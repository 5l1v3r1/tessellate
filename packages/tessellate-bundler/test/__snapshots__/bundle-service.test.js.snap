// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle-service should make a bundle from a source string 1`] = `
"(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory(require(\\"react\\"), require(\\"react-dom\\"));
	else if(typeof define === 'function' && define.amd)
		define([\\"react\\", \\"react-dom\\"], factory);
	else {
		var a = typeof exports === 'object' ? factory(require(\\"react\\"), require(\\"react-dom\\")) : factory(root[\\"react\\"], root[\\"react-dom\\"]);
		for(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];
	}
})(this, function(__WEBPACK_EXTERNAL_MODULE_0__, __WEBPACK_EXTERNAL_MODULE_1__) {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId])
/******/ 			return installedModules[moduleId].exports;
/******/
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// identity function for calling harmony imports with the correct context
/******/ 	__webpack_require__.i = function(value) { return value; };
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 2);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports) {

module.exports = __WEBPACK_EXTERNAL_MODULE_0__;

/***/ }),
/* 1 */
/***/ (function(module, exports) {

module.exports = __WEBPACK_EXTERNAL_MODULE_1__;

/***/ }),
/* 2 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
Object.defineProperty(__webpack_exports__, \\"__esModule\\", { value: true });
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_react__ = __webpack_require__(0);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_0_react___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_0_react__);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_react_dom__ = __webpack_require__(1);
/* harmony import */ var __WEBPACK_IMPORTED_MODULE_1_react_dom___default = __webpack_require__.n(__WEBPACK_IMPORTED_MODULE_1_react_dom__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, \\"Fragment\\", function() { return Fragment; });
/* harmony export (immutable) */ __webpack_exports__[\\"default\\"] = render;
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError(\\"this hasn't been initialised - super() hasn't been called\\"); } return call && (typeof call === \\"object\\" || typeof call === \\"function\\") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== \\"function\\" && superClass !== null) { throw new TypeError(\\"Super expression must either be null or a function, not \\" + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }




var Fragment = function (_Component) {
  _inherits(Fragment, _Component);

  function Fragment() {
    _classCallCheck(this, Fragment);

    return _possibleConstructorReturn(this, (Fragment.__proto__ || Object.getPrototypeOf(Fragment)).apply(this, arguments));
  }

  _createClass(Fragment, [{
    key: 'render',
    value: function render() {
      return __WEBPACK_IMPORTED_MODULE_0_react___default.a.createElement(
        'h1',
        null,
        'Welcome to ',
        this.props.requestURI,
        '!'
      );
    }
  }]);

  return Fragment;
}(__WEBPACK_IMPORTED_MODULE_0_react__[\\"Component\\"]);

function render(element) {
  var props = JSON.parse(element.getAttribute('data-props'));
  __WEBPACK_IMPORTED_MODULE_1_react_dom___default.a.render(__WEBPACK_IMPORTED_MODULE_0_react___default.a.createElement(Fragment, props), element);
}

/***/ })
/******/ ]);
});"
`;
